superset/connectors/sqla/utils.py

```diff
@@ -147,6 +149,24 @@ def get_virtual_table_metadata(dataset: "SqlaTable") -> List[ResultSetColumnType
     return cols
 
 
+def get_columns_description(
+    database: Database,
+    query: str,
+) -> List[ResultSetColumnType]:
+    db_engine_spec = database.db_engine_spec
+    try:
+        with closing(database.get_sqla_engine().raw_connection()) as conn:
+            cursor = conn.cursor()
+            query = database.apply_limit_to_sql(query, limit=1)
+            cursor.execute(query)
+            db_engine_spec.execute(cursor, query)
+            result = db_engine_spec.fetch_data(cursor, limit=1)
+            result_set = SupersetResultSet(result, cursor.description, db_engine_spec)
+            return result_set.columns
+    except Exception as ex:
+        raise SupersetGenericDBErrorException(message=str(ex)) from ex
```

Comment:
I've been thinking about this for quite some time, and one way of doing this would be adding a new endpoint to the dataset API, like `[GET] /api/v1/dataset/{pk}/coltype`, that you could send any arbitrary expression to, and it would return the column metadata for it (the `is_dttm` etc that we now get back in the description). Then we could add a button to the adhoc control that makes it possible to fetch the metadata, and if it's temporal, it would expose the time grain dropdown in the popover.

This would make it possible to have different time grains for different columns, and reuse the same logic here (when we still don't have the frontend designs, we could just call the relevant logic from the backend without an API call).

Reply:
nice point, I will add this detector API in the separated PR.